### PR TITLE
Add methods to get relative paths (#621)

### PR DESCRIPTION
**Pull request by ErikSchierboom**
This PR implements #621 and adds methods to get relative paths for both the `FilePath` and `DirectoryPath` classes.

This is a WIP as I was wondering how you'd want me to do the refactoring. At the moment, there is a lot of duplication, but I was not sure what would be the best way to refactor things. I could move the methods to the abstract base class and add some extract abstract methods, but I'm not sure you'd want to go that way. Any ideas?

### CODE REVIEW

**Comment by patriksvensson** [comment]
Great job! :+1: 

I think that all permutations of this only uses things found in `Cake.Core.IO.Path`, so I guess you could move the code to something like `string GetRelativePath(Path from, Path to)` and return the path as a string. Then you can do the conversion to the right type in each method.

Not sure if this works since I see some minor differences in the code, but maybe that could be solved somehow.

The reason the tests fail is that Mono recognizes `C:/` as a relative path. You can fix this by either converting paths in the tests to Unix paths (without a drive) or mark the tests with a `WindowsTheory`/`WindowsFact` attribute. I think it would be useful to test both Windows and Unix paths here though.

**Comment by ErikSchierboom** [comment]
The PR has been updated. I have now separated the tests into Windows path format tests and Unix format tests. That should fix the Travis build errors.

The code has also been refactored, to remove duplication.

File: src/Cake.Core/IO/DirectoryPath.cs [hunk 1]
```diff
@@ -151,5 +151,70 @@ public static DirectoryPath FromString(string path)
         {
             return new DirectoryPath(path);
         }
+
+        /// <summary>
+        /// Get the relative path to another directory.
+        /// </summary>
+        /// <param name="to">The target directory path.</param>
+        /// <returns>A <see cref="DirectoryPath"/>.</returns>
+        public DirectoryPath GetRelativePath(DirectoryPath to)
```
Comment (patriksvensson) [inline src/Cake.Core/IO/DirectoryPath.cs]: Could we break the logic of this method to it's own file. Mostly an aesthetic adjustment, but would follow how we've done with the PathCollapser stuff.

**Comment by patriksvensson** [comment]
@ErikSchierboom Looks really good! :+1: Left a comment about a minor thing for you.

**Comment by ErikSchierboom** [comment]
@patriksvensson Nice suggestion. I will get on it.

When looking at the `PathCollapserTests` file, I saw some `#if !UNIX` statements. Shouldn't those be converted to `[WindowsFact]` calls?

**Comment by ErikSchierboom** [comment]
@patriksvensson The PR has been updated. I'm not sure about the name though.

**Comment by patriksvensson** [comment]
@ErikSchierboom Maybe call it `RelativePathResolver` with a `Resolve` method. The current name is not bad per se, but might be confused with something that inherits from `Path`. What do you think?

Great finding with the `#if !UNIX` pre-processor. Should absolutely be a `WindowsTheory`, but I think that is better to fix that outside of this PR.

**Comment by ErikSchierboom** [comment]
@patriksvensson Great name! I have updated the PR.

I will convert the `#if !UNIX` calls in a separate PR.

**Comment by patriksvensson** [comment]
@ErikSchierboom Looks good to me. Will merge later tonight once I'm back home. :+1:

**Comment by patriksvensson** [comment]
@ErikSchierboom Is this PR done? I saw the `[WIP]` tag in the title and just want to be sure I don't merge something that you're still working on.

**Comment by ErikSchierboom** [comment]
@patriksvensson Whoops, I totally forgot to remove that. It is ready.

**Comment by patriksvensson** [comment]
@ErikSchierboom Merged! Thanks! :+1:

**Pull request merged**